compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs

```diff
@@ -2145,9 +2153,55 @@ impl<'tcx> InferCtxtPrivExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
             .map(|(_, normalized)| normalized)
             .collect::<Vec<_>>();
 
+        self.suggest_borrowing_to_use_the_candidates(
+            &normalized_impl_candidates,
+            trait_ref,
+            cause.code(),
+            err,
+        );
+
         report(normalized_impl_candidates, err)
     }
 
+    /// Suggests borrowing if there exists an impl candidate that equals to target
+    /// `trait_ref` except `trait_ref`'s self type being a reference to that type.
+    /// Although it's not specific, this method targets `.into()` conversions.
+    fn suggest_borrowing_to_use_the_candidates(
+        &self,
+        impl_candidates: &[TraitRef<'tcx>],
+        trait_ref: ty::PolyTraitRef<'tcx>,
+        code: &ObligationCauseCode<'tcx>,
+        err: &mut Diagnostic,
+    ) {
+        let ty::Ref(_, ty, mutability) = trait_ref.self_ty().skip_binder().kind() else { return };
+        // first arg is the self type and the rest is the arguments for the trait
+        let rest_args = &trait_ref.skip_binder().substs[1..];
+
```

Comment:
instead of doing this in `report_similar_impl_candidates`, can you implement a separate function which uses `infcx.predicate_must_hold_modulo_regions`.

Pretty much, provide that function the `param_env` of the obligation for which we emit the error, look at the self type for whether it is a `Ref(_, ty, mutablity)` and then check for
```rust
self.infcx.predicate_must_hold_modulo_regions(Obligation::new(
    self.tcx,
    ObligtionCause::dummy(),
    param_env,
    trait_ref.with_self_ty(ty),
);
```